src/main/java/org/jenkinsci/test/acceptance/docker/fixtures/GitContainer.java

```diff
@@ -53,5 +53,8 @@ public void addSSHCertificate(String pubKey) throws IOException, InterruptedExce
                 .popen()
                 .verifyOrDieWith("Unable to add SSH public key to authorized keys");
     }
-
+    
+    public static String addBracketsIfNeeded(String ipAddress) {
+        return ipv6Enabled() && !ipAddress.contains("[") ? String.format("[%s]", ipAddress) : ipAddress;
+    }
```

Comment:
`return new URL("http", ipAddress, 666, "").getHost()`?

you do not really want to care if ipv6 is enabled or not - or if you have a DNS name or an IPAddress..

Reply:
> Not following your proposal here, this method is adding brackets to an IPAddress if IPv6 is enabled.

it is adding square brackets **only* if ipv6Enabled() returns true, which is only if you are running dual stack and have asked for this to be enabled.

if you are running on an IPV6 only machine (not dual stack) then you will get an ipv6 address here - and then you are not escaping it at all.

what I am saying is you should not care about the enabling of ipv6 or not - if you use a URL to do the parsing then you will get an escaped host if needed in all situations (dual stack and preferring ipv6, single ipv6 stack etc...).
it will also not escape something if a hostname was returned rather than an IPAddress.

Reply:
Agree, but this is the first attempt to fix this on dual-stack mode, if not applied then it hits 
```
Cloning the remote Git repository
Cloning repository ssh:///git@2600:1900:4020:cd9b::6:22/home/git/gitRepo
 > git init /root/ath/target/jenkinshome11706254101610635272/workspace/busy_analyst # timeout=10
Fetching upstream changes from ssh:///git@2600:1900:4020:cd9b::6:22/home/git/gitRepo
 > git --version # timeout=10
 > git --version # 'git version 2.34.1'
 > git fetch --tags --force --progress -- ssh:///git@2600:1900:4020:cd9b::6:22/home/git/gitRepo +refs/heads/*:refs/remotes/origin/* # timeout=10
ERROR: Error cloning remote repo 'origin'
hudson.plugins.git.GitException: Command "git fetch --tags --force --progress -- ssh:///git@2600:1900:4020:cd9b::6:22/home/git/gitRepo +refs/heads/*:refs/remotes/origin/*" returned status code 128:
stdout: 
stderr: ssh: Could not resolve hostname : Name or service not known
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.

 at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:2842)
 at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandWithCredentials(CliGitAPIImpl.java:2185)
 at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$1.execute(CliGitAPIImpl.java:635)
 at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$2.execute(CliGitAPIImpl.java:871)
 at hudson.plugins.git.GitSCM.retrieveChanges(GitSCM.java:1222)
 at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1305)
 at hudson.scm.SCM.checkout(SCM.java:540)
 at hudson.model.AbstractProject.checkout(AbstractProject.java:1248)
 at hudson.model.AbstractBuild$AbstractBuildExecution.defaultCheckout(AbstractBuild.java:649)
 at jenkins.scm.SCMCheckoutStrategy.checkout(SCMCheckoutStrategy.java:85)
 at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:521)
 at hudson.model.Run.execute(Run.java:1841)
 at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:44)
 at hudson.model.ResourceController.execute(ResourceController.java:101)
 at hudson.model.Executor.run(Executor.java:442)
ERROR: Error cloning remote repo 'origin'
```

Reply:
right - but if you use my suggestion instead it should work regardless of dual stack single stack ipv6/v4 and regardless of having to specify some arbitrary flag.

Reply:
But `GitContainer` is using `ipBound` to retrieve the host info, so at the end it is using the IP (not hostname), same for `LdapContainer`

Reply:
I think I am not getting my point across well.

`ipv6Enabled` is a misnomer - it is not that IPV6 is enabled or not - only that java was started with the system property to prefer ipv6 addresses.
this ONLY makes a difference on dual stac


If you run this code on pure ipv6 stack that preference AFAICT will be `false` and then this escaping will not occur, and this code will then fail.

The code should not care if it is preferring ipv6 or ipv4.  All it should care about is if `ipAddress` is an ipv6 address (not already enclosed in `[` ) then to escape it.

And helpfully there is some code that can do this for you, it is in `URL` and pointed out at the start of this thread.  If it is passed an ipv4 literal address it will get an opv4 litteral address, pass it an ipv6 litteral and you get back the litteral address enclosed in `[]`, pass it something already enclosed and you get that back, pass it anything else like a string and you get that back to.

Reply:
> ipv6Enabled is a misnomer - it is not that IPV6 is enabled or not - only that java was started with the system property to prefer ipv6 addresses. This ONLY makes a difference on dual stack

Then the issue is on https://github.com/jenkinsci/docker-fixtures/blob/master/src/main/java/org/jenkinsci/test/acceptance/docker/DockerContainer.java#L239-L241 not in this PR.

> If you run this code on pure ipv6 stack that preference AFAICT will be false and then this escaping will not occur, and this code will then fail.

Agree, the same if you run it as it is on dual stack before this PR. With this PR it will work (on dual stack)

Reply:
> Agree, the same if you run it as it is on dual stack before this PR. With this PR it will work (on dual stack)

but with the suggestion it would work everywhere! and be simpler code!

Reply:
My understanding from James comment is that instead of 

`ssh://git@" + addBracketsIfNeeded(host()) + ":" + port() + REPO_DIR` 

You build an URL object which can take a bracketed or non bracketed ipv6 ip as `host` and then use the `toExternalForm()` method.

This way the ATH do not care at all about the stack and will work everywhere

Reply:
FTR I am not opposed to a function that uses `new URL(....).getHost()`  rather than inlining it (to make the code more understandable)

Reply:
close but `toExternalForm` will give you a port and scheme, which is not what you need and then ned some parsing.  calling `getHost()` is enough to get the escaped (if required) host after construction.

Reply:
Updated in https://github.com/jenkinsci/acceptance-test-harness/pull/1510/commits/5338f366038f3de2105851ac1ff39d83064583a8 to improve readability

Reply:
> close but toExternalForm will give you a port and scheme, which is not what you need and would then need some parsing to get the (possibly escaped) host.

I mean for existing code like https://github.com/jenkinsci/acceptance-test-harness/blob/master/src/main/java/org/jenkinsci/test/acceptance/docker/fixtures/GitContainer.java#L30-L37 which is basically constructing the external form of an url by using strings...